static/js/csrf.ts

```diff
@@ -11,7 +11,11 @@ $(() => {
     }
 
     $.ajaxSetup({
-        beforeSend(xhr, settings) {
+        beforeSend(xhr: JQuery.jqXHR, settings: JQuery.AjaxSettings) {
+            if (!settings.url || csrf_token === undefined) {
+                return;
+            }
```

Comment:
I wonder for things like this if we want to instead be writing an `assert` type statement -- basically if we consider it a bug in the code to call this without those properties set.

Reply:
I don't think it is a bug if this function is called without those properties set for two reasons. The first is that the `csrf_token` is never going to be `undefined`  because we check that it is not undefined before we set up this `beforeSend` handler and never modify it afterward. TypeScript intentionally makes us double-check here because it _can be_ modified outside this handler. For the `settings.url` case, I don't know why it would be undefined but, in any case, I think the handler is supposed to be called for any URL we request for so this is fine.

Reply:
OK; maybe worth a comment noting that both of these situations should never occur?

Reply:
There seems to have been a miscommunication here. If our understanding is that some condition should not occur, that means it *is* a bug for that condition to occur, and so we should throw an error if we encounter that condition.

Also, since `settings.url` is `string | undefined`, testing it for truthiness is ambiguous; it makes it unclear whether you’re testing for `""` or for `undefined`. Better to be explicit.

Reply:
OK, I added a `blueslip.error` here.

Reply:
I think we want `throw new Error` here instead, since that prevents executing the following code (which can't work without these properties).

`blueslip.error` just logs an error but proceeds from there as normal.

Reply:
Done.